# Introduction and Data Wrangling

A fictitious US shopping mall named "Blue Pocket" aims to evaluate their marketing campaigns' performance using 5 months of data from 2021. Their promotions include paid search coupons and promo codes.

**Dataset Dictionary:**

- **Ad Group:** Category of the advertisement (e.g., coupon/promo code, desktop ad/mobile ad).
- **Month:** Campaign month. The campaign ran from July 2021 to November 2021.
- **Impressions:** Digital marketing metric quantifying the number of views or engagements an advertisement receives, also known as an **ad view**.
- **Clicks:** Number of clicks the ad received.
- **CTR:** Click-Through Rate—the number of clicks an ad receives divided by the number of times it's shown (**clicks ÷ impressions = CTR**).
- **Conversions:** Valuable actions users take on a website, such as making a purchase or filling out a form. Success is measured by the number of conversions generated at a **particular cost**.
- **Conv Rate:** Conversion Rate—the percentage of people who convert after clicking on ads. Conversions may include making a purchase, completing a contact form, or requesting a **free trial**, depending on campaign goals.
- **Cost:** Actual money spent by the advertiser (the "shop") for the related ad group.
- **CPC:** Cost Per Click—the cost of specific ads divided by the number of clicks. This metric evaluates campaign effectiveness in terms of ROI (Return on Investment). A lower or decreasing CPC is preferable to a higher or increasing one.
- **Revenue:** Total income generated by the advertisement.
- **Sale Amount:** Quantity of sales derived from a single ad group.
- **P&L:** Profit and Loss, calculated using the formula **Revenue - Cost**.

>The Blue Pocket Shopping Mall has been promoting its website for both mobile and desktop platforms. To evaluate campaign performance, focus on conversion rate, profit, and ROI—key metrics that drive sales growth. To understand fluctuations in conversion rate, analyze CTR. Additionally, examining impressions will provide insights into customer behavior, particularly which search keywords or phrases attracted the most visitors and how that attraction translated into purchases. For managing paid search campaigns, prioritize CPC to find the most cost-effective keywords for attracting visitors from an ROI perspective. Also, consider the cost per campaign to optimize budget allocation.

# I. Data Inspection

In [38]:
# Impot labraries
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

In [39]:
# Load Data source
DataLink = ('https://raw.githubusercontent.com/Evank2023/Portfolio/refs/heads/ShoppingMallPaidSearchCampaign/final_shop_6modata.csv')
Campaign_Data = pd.read_csv(DataLink)

---

In [40]:
Campaign_Data.head()

Ad Group Month  Impressions  Clicks   CTR  \
0    Shop - 1:1 - Desk - [shop coupon code]  July        16038    6504  0.41   
1         Shop - 1:1 - Desk - [shop coupon]  July        36462   14367  0.39   
2  Shop - 1:1 - Desk - [shop discount code]  July         3635    1458  0.40   
3     Shop - 1:1 - Desk - [shop promo code]  July        26185   10418  0.40   
4          Shop - 1:1 - Desk - [shop promo]  July          808     282  0.35   

   Conversions  Conv Rate   Cost   CPC  Revenue  Sale Amount      P&L  
0         1166       0.10   6669  1.03     6402    136770.05 -267.086  
1         2188       0.09  13746  0.96    13262    283215.21 -483.951  
2          248       0.09   1606  1.10     1723     39165.46  117.136  
3         2294       0.12  13278  1.27    13042    284823.48 -235.921  
4           61       0.15    391  1.39      337      7717.77  -53.604

In [41]:
Campaign_Data.tail()

Ad Group     Month  Impressions  \
185  Shop - Exact - Desk - Black Friday/Cyber Monday  November          257   
186         Shop - 1:1 - Desk - [shop discount code]  November         7254   
187                Shop - Exact - Desk - Coupon Code  November        18526   
188   Shop - Exact - Mob - Black Friday/Cyber Monday  November         3662   
189                 Shop - Exact - Desk - Promo Code  November        25592   

     Clicks   CTR  Conversions  Conv Rate   Cost   CPC  Revenue  Sale Amount  \
185      24  0.09            7       0.28      3  0.14       45       898.80   
186    2725  0.38          512       0.11   3182  1.17     3227     66672.29   
187    5553  0.30          919       0.10   5982  1.08     6047    129556.90   
188     266  0.07           24       0.09     44  0.17      160      3268.63   
189    7726  0.30         1731       0.14  10914  1.41    11223    236665.59   

         P&L  
185   41.946  
186   45.468  
187   64.552  
188  115.963  
189  308.975

In [42]:
Campaign_Data.columns

Index(['Ad Group', 'Month', 'Impressions', 'Clicks', 'CTR', 'Conversions',
       'Conv Rate', 'Cost', 'CPC', 'Revenue', 'Sale Amount', 'P&L'],
      dtype='object')

In [43]:
Campaign_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Ad Group     190 non-null    object 
 1   Month        190 non-null    object 
 2   Impressions  190 non-null    int64  
 3   Clicks       190 non-null    int64  
 4   CTR          190 non-null    float64
 5   Conversions  190 non-null    int64  
 6   Conv Rate    190 non-null    float64
 7   Cost         190 non-null    int64  
 8   CPC          190 non-null    float64
 9   Revenue      190 non-null    int64  
 10  Sale Amount  190 non-null    float64
 11  P&L          190 non-null    float64
dtypes: float64(5), int64(5), object(2)
memory usage: 17.9+ KB


In [44]:
Campaign_Data.describe()

Impressions        Clicks         CTR  Conversions   Conv Rate  \
count     190.000000    190.000000  190.000000   190.000000  190.000000   
mean    14077.363158   4865.805263    0.272105   505.242105    0.079737   
std     29771.686227  11348.529219    0.107894  1052.202922    0.052859   
min        35.000000      2.000000    0.050000     0.000000    0.000000   
25%      1065.000000    264.500000    0.192500    24.000000    0.050000   
50%      4969.000000    930.000000    0.285000    70.500000    0.070000   
75%     13380.000000   4190.500000    0.360000   428.250000    0.100000   
max    276568.000000  99526.000000    0.470000  7563.000000    0.500000   

               Cost         CPC       Revenue    Sale Amount          P&L  
count    190.000000  190.000000    190.000000     190.000000   190.000000  
mean    3344.063158    0.791263   2957.684211   63416.180579  -386.361916  
std     6524.606753    0.403312   5962.413097  125414.656922   903.073776  
min        1.000000    0.140000      0.000000       0.000000 -5672.271000  
25%      188.500000    0.480000    144.500000    2985.135000  -294.977500  
50%      563.500000    0.635000    459.500000   10274.105000   -75.978000  
75%     2967.000000    1.107500   2672.250000   59345.195000    -6.027500  
max    43542.000000    2.020000  42440.000000  886095.310000  1507.685000

> According to the info() function, there are **no missing values**, and data types correspond appropriately.\
> **In summary**, across 190 rows of data, there's a wide range between minimum and maximum values, indicating excessively underperforming campaigns during the 5-month period.\
> Analyzing each minimum and maximum value would clarify which campaigns performed best or worst based on different criteria.\
> The analysis results would pinpoint precise areas for developing targeted strategies to improve future campaign performance.

---

# II. Break down the 'Ad Group' column to understand the types of promotions and how they were organized.

In [25]:
# How many type of promotions are there ?

AD_Grooup = Campaign_Data['Ad Group'].unique()
AD_Group_Unique_df = pd.DataFrame(AD_Grooup, columns=['Ad Groups'])
AD_Group_Unique_df

Ad Groups
0            Shop - 1:1 - Desk - [shop coupon code]
1                 Shop - 1:1 - Desk - [shop coupon]
2          Shop - 1:1 - Desk - [shop discount code]
3             Shop - 1:1 - Desk - [shop promo code]
4                  Shop - 1:1 - Desk - [shop promo]
5             Shop - 1:1 - Mob - [shop coupon code]
6                  Shop - 1:1 - Mob - [shop coupon]
7           Shop - 1:1 - Mob - [shop discount code]
8              Shop - 1:1 - Mob - [shop promo code]
9                   Shop - 1:1 - Mob - [shop promo]
10                 Shop - Exact - Desk - Competitor
11                Shop - Exact - Desk - Coupon Code
12              Shop - Exact - Desk - Discount Code
13              Shop - Exact - Desk - Free Shipping
14                      Shop - Exact - Desk - Offer
15                 Shop - Exact - Desk - Promo Code
16                       Shop - Exact - Desk - Sale
17                  Shop - Exact - Mob - Competitor
18                 Shop - Exact - Mob - Coupon Code
19               Shop - Exact - Mob - Discount Code
20               Shop - Exact - Mob - Free Shipping
21                       Shop - Exact - Mob - Offer
22                  Shop - Exact - Mob - Promo Code
23                        Shop - Exact - Mob - Sale
24                Shop - Phrase - Desk - Competitor
25               Shop - Phrase - Desk - Coupon Code
26             Shop - Phrase - Desk - Discount Code
27             Shop - Phrase - Desk - Free Shipping
28                     Shop - Phrase - Desk - Offer
29                Shop - Phrase - Desk - Promo Code
30                      Shop - Phrase - Desk - Sale
31                 Shop - Phrase - Mob - Competitor
32                Shop - Phrase - Mob - Coupon Code
33              Shop - Phrase - Mob - Discount Code
34              Shop - Phrase - Mob - Free Shipping
35                      Shop - Phrase - Mob - Offer
36                 Shop - Phrase - Mob - Promo Code
37                       Shop - Phrase - Mob - Sale
38  Shop - Exact - Desk - Black Friday/Cyber Monday
39   Shop - Exact - Mob - Black Friday/Cyber Monday

In [28]:
# Split the 'Ad Groups' column into multiple columns to create a new DataFrame

AD_Group_Split = AD_Group_Unique_df['Ad Groups'].str.split(' - ', expand=True)
AD_Group_Split.columns = ['Place', 'Target', 'Platform', 'Promotion']
AD_Group_Split

Place  Target Platform                  Promotion
0   Shop     1:1     Desk         [shop coupon code]
1   Shop     1:1     Desk              [shop coupon]
2   Shop     1:1     Desk       [shop discount code]
3   Shop     1:1     Desk          [shop promo code]
4   Shop     1:1     Desk               [shop promo]
5   Shop     1:1      Mob         [shop coupon code]
6   Shop     1:1      Mob              [shop coupon]
7   Shop     1:1      Mob       [shop discount code]
8   Shop     1:1      Mob          [shop promo code]
9   Shop     1:1      Mob               [shop promo]
10  Shop   Exact     Desk                 Competitor
11  Shop   Exact     Desk                Coupon Code
12  Shop   Exact     Desk              Discount Code
13  Shop   Exact     Desk              Free Shipping
14  Shop   Exact     Desk                      Offer
15  Shop   Exact     Desk                 Promo Code
16  Shop   Exact     Desk                       Sale
17  Shop   Exact      Mob                 Competitor
18  Shop   Exact      Mob                Coupon Code
19  Shop   Exact      Mob              Discount Code
20  Shop   Exact      Mob              Free Shipping
21  Shop   Exact      Mob                      Offer
22  Shop   Exact      Mob                 Promo Code
23  Shop   Exact      Mob                       Sale
24  Shop  Phrase     Desk                 Competitor
25  Shop  Phrase     Desk                Coupon Code
26  Shop  Phrase     Desk              Discount Code
27  Shop  Phrase     Desk              Free Shipping
28  Shop  Phrase     Desk                      Offer
29  Shop  Phrase     Desk                 Promo Code
30  Shop  Phrase     Desk                       Sale
31  Shop  Phrase      Mob                 Competitor
32  Shop  Phrase      Mob                Coupon Code
33  Shop  Phrase      Mob              Discount Code
34  Shop  Phrase      Mob              Free Shipping
35  Shop  Phrase      Mob                      Offer
36  Shop  Phrase      Mob                 Promo Code
37  Shop  Phrase      Mob                       Sale
38  Shop   Exact     Desk  Black Friday/Cyber Monday
39  Shop   Exact      Mob  Black Friday/Cyber Monday

In [31]:
# Simplify splited data frame extracting unique values from each column in the AD_Group_Split DataFrame
unique_place = AD_Group_Split['Place'].unique()
unique_target = AD_Group_Split['Target'].unique()
unique_platform = AD_Group_Split['Platform'].unique()
unique_promotion = AD_Group_Split['Promotion'].unique()

# Create a dictionary to store the extracted values
unique_values_dict = {
    'Place': unique_place,
    'Target': unique_target,
    'Platform': unique_platform,
    'Promotion': unique_promotion
}

# Convert the dictionary into a DataFrame
unique_values_df = pd.DataFrame.from_dict(unique_values_dict, orient='index').transpose()

# Show the resulting DataFrame
unique_values_df

   Place  Target Platform                  Promotion
0   Shop     1:1     Desk         [shop coupon code]
1   None   Exact      Mob              [shop coupon]
2   None  Phrase     None       [shop discount code]
3   None    None     None          [shop promo code]
4   None    None     None               [shop promo]
5   None    None     None                 Competitor
6   None    None     None                Coupon Code
7   None    None     None              Discount Code
8   None    None     None              Free Shipping
9   None    None     None                      Offer
10  None    None     None                 Promo Code
11  None    None     None                       Sale
12  None    None     None  Black Friday/Cyber Monday


### Summary of Ad Group Data:
* Campaigns: 40 distinct types
* Promotion Types: 14 categories
* Targeting Strategies: 3 (One-to-one match, Exact keyword matching, Phrase search matching)
* Platforms: Desktop and Mobile
* Promotion Examples: Coupon codes, Discounts, Free Shipping, Competitor offers, Sales, Black FridayCyber Monday deals\
**The paid search campaigns utilize three targeting strategies: one-to-one match, exact keyword matching, and phrase search matching.**

# III. Sales Amount Analysis

# IV. Conversion Analysis via impressions and Click Through

# V. ROI Anlaysis 1 | CPC VS Conversion Rate

# VI. ROI Analysis 2 | Campaign Cost VS Profit and Loss

In [ ]:


# Assuming Campaign_Data is your DataFrame and it has 'Ad Group' and 'P&L' columns
aggregated_df = Campaign_Data.groupby('Ad Group')['P&L'].sum().reset_index()

# Renaming the column for clarity (optional)
aggregated_df.rename(columns={'P&L': 'Total P&L'}, inplace=True)

# Sort the DataFrame by 'Total P&L' in descending order (from highest to lowest)
aggregated_df_sorted = aggregated_df.sort_values(by='Total P&L', ascending=False)

# Display the sorted DataFrame
print(aggregated_df_sorted)


# VII. Conclusion

Which campaign performed best and worst?

What factors influenced the conversion rates?

Did the best-performing campaign generate the most revenue?

In terms of paid search management, how can we promote future campaigns more efficiently, considering cost management to boost profits?

What should we target for future growth?

Although this data doesn't contain exact keywords and phrases, limiting our understanding of how specific keywords or phrases performed, it still reveals that...